### DISCUSSION TITLE
Capture Zigbee OTA firmware URL from Wireshark traffic sniffing

### DISCUSSION THREAD

**Discussion by Hedda** [zigpy/zigpy#723] [General]
FYI @pipiche38 has begun an interesting project that is trying to get OTA firmware URL via info get from Wireshark Zigbee sniffing:

https://github.com/pipiche38/Capture-OTA-from-Wireshark

https://github.com/pipiche38/Domoticz-Zigate-Wiki/blob/master/en-eng/Corner_Retreiving-Legrand-Firmware.md

Update: compujuckel also written a small program for extracting and reassembling Zigbee OTA images file from Wireshark packet captures:

https://github.com/compujuckel/ZigbeeOtaExtractor

Zigbee OTA Extractor - Tool to extract Zigbee OTA images from packet captures

**Comment by tube0013** [comment]
I did this with Puddy's help with a Terncy device.

**Comment by Hedda** [comment]
Do you think having a way to easier get OTA URLs could eventually lead to more OTA providers being added to zigpy by default?

https://github.com/zigpy/zigpy/tree/dev/zigpy/ota

**Reply by puddly** [reply]
Very few (three?) OTA providers have OTA URLs. The vast majority of known images (https://github.com/Koenkk/zigbee-OTA/) are the image files themselves or direct links to them.

pipiche38's approach captures the OTA image as it's being sent by the coordinator to the target device over-the-air. This works well for every device if you put the sniffer in a good location but if you miss even one packet, the captured OTA file is incomplete and since the device likely upgraded, you won't be able to try again.

When zigpy eventually gets support for non-coordinator modes, it'll be good to combine the two and allow zigpy to itself pretend to be the device, connect to the proprietary hub, and just "upgrade" itself to capture the image.

**Reply by Hedda** [reply]
> Very few (three?) OTA providers have OTA URLs. The vast majority of known images (https://github.com/Koenkk/zigbee-OTA/) are the image files themselves or direct links to them.

Wonder if the new zigpy based Jeedom plugin handles OTA provision differently as blog post makes it sound as if they do?

https://blog.jeedom.com/5183-tout-ce-quil-faut-savoir-sur-le-plugin-officiel-zigbee/ (translated from French to English):

_**Updating OTA modules**_

_The Zigbee plugin makes it easy to make the network more reliable by automatically managing module firmware updates . Indeed, after activating the corresponding option in the plugin, the modules will be automatically updated in the presence of new firmware or if a module requests it. It is also possible to update the firmware of a module manually by clicking on the button provided for this purpose._

_Of course, to work, this tool must be able to download new firmware regularly. The manufacturer must therefore communicate this firmware in some way. Before any purchase of Zigbee equipment, we strongly advise you to find out about the correct provision of this firmware by the manufacturer in order to guarantee the sustainability of your installation._

**Comment by MattWestb** [comment]
@puddly I using the device type "Zigbee control bridge" for pairing other networks and can sending and reviving commands but i have not looking of spooking the firmware image ID for one device but ZHA is trying to find if its some new IKEA firmware then i using there MF-ID but not finding any (then its very likely zero).
Perhaps i shall trying putting in on real ID and see wot i happening

**Reply by MattWestb** [reply]
I think its possible cheating most (no tuya)  devices by setting the manufacturer in in Cluster config and then adding the image ID in the OTA plugin.
Was doing on fast test but it was missing many callbacks that im to lazy for fixing.
![WSTK-OTS01](https://user-images.githubusercontent.com/49618193/116711149-dd8a8080-a9d2-11eb-87b1-c4491a8eb095.PNG)

The 4354 is the signal repeater ;-))

With little more work and knowing C++ i think its possible making the manufacture and device ID being sett in the device by CLI only making it as one on/off plug or dimmer light so all GW can joining it and then requesting one OTA image from the original GW. and sniffing its URL or downloading and / or not deleting the OTA storage and dumping the flash with  the OTA file.

**Reply by pipiche38** [reply]
This is indeed the best way to capture the data directly from the source. The Wireshark story is nearly mission impossible

**Reply by MattWestb** [reply]
If having on WSTK (silabs devkit) you can doing one false OTA request and still using wireshark for getting the URL and then you can grabbing the "original file" from the server and dont need sniffing the zigbee packages.
If its not possible grabbing the URL is still possible sniffing the Zigbee packages and building the file that is being sent to the device. And the doing one "pipiche-OTA file" of it :-))

I think some is packing the OTA file on the server so it cant being used but  in the Zigbee side it shall being on version that is accepted of all device with the right manufacture and device ID.

**Reply by Adminiuga** [reply]
you don't need to have a silabs kit. Just need to be a router on the network and send OTA image request with the required image_id, hardware_id, etc.
AFAIK deCONZ has OTA plugin which can do that, but I haven't tried personally.

**Comment by Hedda** [comment]
FYI, compujuckel has written a small program for extracting and reassembling Zigbee OTA images file from Wireshark packet captures:

https://github.com/compujuckel/ZigbeeOtaExtractor

Zigbee OTA Extractor - Tool to extract Zigbee OTA images from packet captures

Read about it in this thread:

https://github.com/Koenkk/zigbee2mqtt/issues/14926

**Reply by MattWestb** [reply]
Our **Puddly**s version is working  great sniffing Dirigera OTAs (have doing 6 updates that TF not having) and dont need extra program installed if having one python with zigpy-CLI installed for testing.
It can also fixing if frames is missing and using more sniffs for getting all frames for the complete OTA file and IKEA signing is not  being broken = 110% safe.
https://github.com/zigpy/zigpy-cli/pull/29
